File: src/services/passwordService.ts
import { auth } from '@/lib/firebase';
import { updatePassword } from 'firebase/auth';
import { doc, updateDoc, Timestamp } from 'firebase/firestore';
import { db } from '@/lib/firebase';
import { generateDefaultPassword } from '@/utils/passwordGenerator';

/**
 * Service for password management operations
 */
export class PasswordService {
  /**
   * Reset a user's password (admin only)
   * Calls the Vercel API endpoint
   *
   * @param userId - The user ID to reset
   * @param clubId - The club ID
   * @param newPassword - Optional custom password (defaults to CalyCompta{year}-{month}-{unique})
   * @param requirePasswordChange - Whether user must change password on next login
   */
  static async resetUserPassword(
    userId: string,
    clubId: string,
    newPassword?: string,
    requirePasswordChange: boolean = true
  ): Promise<void> {
    // Use secure generated password if none provided
    const passwordToSet = newPassword || generateDefaultPassword();

    const currentUser = auth.currentUser;
    if (!currentUser) {
      throw new Error("Vous devez être authentifié");
    }

    const idToken = await currentUser.getIdToken();

    const response = await fetch(`/api/reset-password?t=${Date.now()}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${idToken}`,
        'Cache-Control': 'no-cache, no-store, must-revalidate',
      },
      body: JSON.stringify({
        userId,
        clubId,
        newPassword: passwordToSet,
        requirePasswordChange,
      }),
    });

    const result = await response.json();

    if (!response.ok) {
      throw new Error(result.message || `Erreur ${response.status}`);
    }

    return result;
  }

  /**
   * Change current user's password
   * Also updates requirePasswordChange flag in Firestore
   */
  static async changeMyPassword(
    newPassword: string,
    clubId: string
  ): Promise<void> {
    const currentUser = auth.currentUser;
    if (!currentUser) {
      throw new Error("Vous devez être authentifié");
    }

    // Update password in Firebase Auth
    await updatePassword(currentUser, newPassword);

    // Update requirePasswordChange flag in Firestore
    const userDocRef = doc(db, 'clubs', clubId, 'members', currentUser.uid);
    await updateDoc(userDocRef, {
      requirePasswordChange: false,
      updatedAt: Timestamp.now(),
    });
  }

  /**
   * Check if user needs to change password
   */
  static async checkRequirePasswordChange(
    userId: string,
    clubId: string
  ): Promise<boolean> {
    const { doc: firestoreDoc, getDoc } = await import('firebase/firestore');
    const userDocRef = firestoreDoc(db, 'clubs', clubId, 'members', userId);
    const userDoc = await getDoc(userDocRef);

    if (!userDoc.exists()) {
      return false;
    }

    const userData = userDoc.data();
    return userData?.requirePasswordChange === true;
  }
}

File: lib/services/password_service.dart
import 'dart:convert';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

/// Service for password management
///
/// Provides password change functionality via the Vercel API.
/// Used for both forced first-time password change and voluntary changes.
class PasswordService {
  static const String _baseUrl = 'https://caly.club/api';

  /// Change user's own password via backend API
  ///
  /// This endpoint:
  /// - Updates the password in Firebase Auth
  /// - Clears the requirePasswordChange flag in Firestore
  /// - Invalidates other sessions
  /// - Creates an audit log entry
  ///
  /// Throws [PasswordChangeException] on error
  Future<void> changePassword({
    required String userId,
    required String clubId,
    required String newPassword,
  }) async {
    final user = FirebaseAuth.instance.currentUser;
    if (user == null) {
      throw PasswordChangeException('Utilisateur non connecté');
    }

    try {
      debugPrint('🔐 [PasswordService] Changing password for user: $userId');

      final authToken = await user.getIdToken();

      final response = await http.post(
        Uri.parse('$_baseUrl/change-password'),
        headers: {
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'userId': userId,
          'clubId': clubId,
          'authToken': authToken,
          'newPassword': newPassword,
        }),
      );

      if (response.statusCode == 200) {
        debugPrint('✅ [PasswordService] Password changed successfully');
        return;
      }

      // Handle error responses
      final errorData = jsonDecode(response.body);
      final errorMessage = errorData['error'] ?? 'Erreur inconnue';

      debugPrint('❌ [PasswordService] Error: $errorMessage');

      switch (response.statusCode) {
        case 401:
          throw PasswordChangeException(
            'Session expirée. Veuillez vous reconnecter.',
          );
        case 403:
          throw PasswordChangeException(
            'Vous n\'êtes pas autorisé à effectuer cette action.',
          );
        case 404:
          throw PasswordChangeException(
            'Utilisateur non trouvé.',
          );
        default:
          throw PasswordChangeException(errorMessage);
      }
    } on PasswordChangeException {
      rethrow;
    } catch (e) {
      debugPrint('❌ [PasswordService] Unexpected error: $e');
      throw PasswordChangeException(
        'Erreur lors du changement de mot de passe. Veuillez réessayer.',
      );
    }
  }

  /// Validate password meets requirements
  ///
  /// Requirements:
  /// - At least 8 characters
  /// - At least one uppercase letter
  /// - At least one lowercase letter
  /// - At least one number
  static PasswordValidation validatePassword(String password) {
    final hasMinLength = password.length >= 8;
    final hasUppercase = password.contains(RegExp(r'[A-Z]'));
    final hasLowercase = password.contains(RegExp(r'[a-z]'));
    final hasNumber = password.contains(RegExp(r'[0-9]'));

    return PasswordValidation(
      hasMinLength: hasMinLength,
      hasUppercase: hasUppercase,
      hasLowercase: hasLowercase,
      hasNumber: hasNumber,
    );
  }

  /// Get password requirements text in French
  static String getPasswordRequirementsText() {
    return 'Le mot de passe doit contenir :\n'
        '• Au moins 8 caractères\n'
        '• Une lettre majuscule\n'
        '• Une lettre minuscule\n'
        '• Un chiffre';
  }
}

/// Password validation result
class PasswordValidation {
  final bool hasMinLength;
  final bool hasUppercase;
  final bool hasLowercase;
  final bool hasNumber;

  const PasswordValidation({
    required this.hasMinLength,
    required this.hasUppercase,
    required this.hasLowercase,
    required this.hasNumber,
  });

  /// Returns true if all requirements are met
  bool get isValid => hasMinLength && hasUppercase && hasLowercase && hasNumber;

  /// Returns the number of requirements met
  int get metCount =>
      (hasMinLength ? 1 : 0) +
      (hasUppercase ? 1 : 0) +
      (hasLowercase ? 1 : 0) +
      (hasNumber ? 1 : 0);
}

/// Exception thrown when password change fails
class PasswordChangeException implements Exception {
  final String message;

  PasswordChangeException(this.message);

  @override
  String toString() => message;
}
